webapp/src/js/bootstrapper/initial-replication.js

```diff
@@ -0,0 +1,188 @@
+    $('#btn-abort').click(() => {
+      document.cookie = 'login=force;path=/';
+      window.location.reload(false);
+    });
+  });
+};
+
+const getMissingDocIdsRevsPairs = async (localDb, remoteDocIdsRevs) => {
+  const localDocs = await getLocalDocList(localDb);
+  return remoteDocIdsRevs.filter(({ id, rev }) => !localDocs[id] || localDocs[id] !== rev);
+};
+
+const getDownloadList = async (localDb) => {
+  const response = await utils.fetchJSON('/api/v1/initial-replication/get-ids');
+
+  docIdsRevs = await getMissingDocIdsRevsPairs(localDb, response.doc_ids_revs);
+  lastSeq = response.last_seq;
+  remoteDocCount = response.doc_ids_revs.length;
+
+  if (response.warn) {
+    await displayTooManyDocsWarning(response);
+  }
+};
+
+const getLocalDocList = async (localDb) => {
+  const response = await localDb.allDocs();
+
+  const localDocMap = {};
+  response.rows.forEach(row => localDocMap[row.id] = row.value && row.value.rev);
+  return localDocMap;
+};
+
+const getDocsBatch = async (remoteDb, localDb) => {
+  const batch = docIdsRevs.splice(0, BATCH_SIZE);
+
+  if (!batch.length) {
+    return;
+  }
+
+  const res = await remoteDb.bulkGet({ docs: batch, attachments: true });
+  const docs = res.results
+    .map(result => result.docs && result.docs[0] && result.docs[0].ok)
+    .filter(doc => doc);
+  await localDb.bulkDocs(docs, { new_edits: false });
+};
+
+const downloadDocs = async (remoteDb, localDb) => {
+  setUiStatus('FETCH_INFO', { count: remoteDocCount - docIdsRevs.length, total: remoteDocCount });
+  do {
+    await getDocsBatch(remoteDb, localDb);
+    setUiStatus('FETCH_INFO', { count: remoteDocCount - docIdsRevs.length, total: remoteDocCount });
+  } while (docIdsRevs.length > 0);
+};
+
+const writeCheckpointers = async (remoteDb, localDb) => {
+  const replicateFromPromise = localDb.replicate.from(remoteDb, {
+    live: false,
+    retry: false,
+    heartbeat: 10000,
+    timeout: 1000 * 60 * 10, // try for ten minutes then give up,
+    query_params: { initial_replication: true },
+    since: lastSeq,
+  });
+
+  replicateFromPromise.on('change', () => replicateFromPromise.cancel());
+  await replicateFromPromise;
+
+  const localInfo = await localDb.info();
+  await localDb.replicate.to(remoteDb, {
+    since: localInfo.update_seq,
+  });
+};
+
+const replicate = async (remoteDb, localDb) => {
+  setUiStatus('LOAD_APP');
+
+  await startInitialReplication(localDb);
+
+  setUiStatus('POLL_REPLICATION');
+  await getDownloadList(localDb);
+  await downloadDocs(remoteDb, localDb);
+  await writeCheckpointers(remoteDb, localDb);
+
+  await completeInitialReplication(localDb);
+};
+
+const getReplicationLog = async (localDb) => {
+  try {
+    return await localDb.get(INITIAL_REPLICATION_LOG);
+  } catch (err) {
+    return null;
+  }
+};
+const isReplicationNeeded = async (localDb, userCtx) => {
+  const requiredDocs = [
+    '_design/medic-client',
+    'settings',
+    `org.couchdb.user:${userCtx.name}`,
```

Comment:
this doc wasn't previously required right? seems like a good requirement
i'm curious if you're being defensive here or if you are fixing any known problem

Reply:
i also wonder what happens if initial replication fails (for whatever reason). 
when the user retries they will get the slow experience if these few docs were successful
why don't we put them into this 10x faster experience until initial replication actually completes?

Reply:
> if you are fixing any known problem

I would not call it known - because there is no associated issue afaik, but it is a problem old initial replication had, where a user's app could start up without the user-settings doc and have all sorts of issues - like failing to load tasks for example. 

I'll make some time this week to simulate and create the issue, just so it's logged somewhere.